curvefs/src/space/metaserver_client.cpp

```diff
@@ -0,0 +1,133 @@
+/*
+ *  Copyright (c) 2021 NetEase Inc.
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+#include "curvefs/src/space/metaserver_client.h"
+
+#include <glog/logging.h>
+
+#include <algorithm>
+
+namespace curvefs {
+namespace space {
+
+DEFINE_string(metaServerAddr, "0.0.0.0:20000", "metaserver service address");
+
+bool MetaServerClient::Init(const MetaServerClientOption& opt) {
+    opt_ = opt;
+
+    if (opt_.addr.empty()) {
+        opt_.addr = FLAGS_metaServerAddr;
+    }
+
+    int ret = channel_.Init(opt_.addr.c_str(), nullptr);
+    if (ret != 0) {
+        LOG(ERROR) << "Init channel to metaserver failed";
+        return false;
+    }
+
+    return true;
+}
+
+bool MetaServerClient::GetAllInodeExtents(uint32_t fsId, uint64_t rootInodeId,
+                                          Extents* exts) {
+    return RecursiveListDentry(fsId, rootInodeId, exts, nullptr);
+}
+
+bool MetaServerClient::GetMaxChunkIdOfAllInode(uint32_t fsId,
+                                               uint64_t rootInodeId,
+                                               uint64_t* maxChunkId) {
+    return RecursiveListDentry(fsId, rootInodeId, nullptr, maxChunkId);
+}
+
+// TODO(wuhanqing): Using recursion, if the file system has a deep layer, will
+// blow up the stack?
+bool MetaServerClient::RecursiveListDentry(uint32_t fsId, uint64_t inodeId,
+                                           Extents* exts,
+                                           uint64_t* maxChunkId) {
+    if (exts == nullptr && maxChunkId == nullptr) {
+        return false;
+    } else if (exts != nullptr && maxChunkId != nullptr) {
+        return false;
+    }
+
+    metaserver::MetaServerService_Stub stub(&channel_);
+    metaserver::ListDentryRequest request;
+    metaserver::ListDentryResponse response;
+
+    brpc::Controller cntl;
+
+    request.set_fsid(fsId);
+    request.set_dirinodeid(inodeId);
+    stub.ListDentry(&cntl, &request, &response, nullptr);
+
+    if (cntl.Failed() || response.statuscode() != metaserver::OK) {
+        return false;
+    }
+
+    for (auto& d : response.dentrys()) {
```

Comment:
add a recursion limit will better

Reply:
already added a todo, will fix this later